Fix shellcheck errors in scripts (part 1)

MuteSpirit opened the pull request:
scripts: fix shellcheck issues in tools/docker/image-pull.sh

Will contain one commit only after merge #1103

Commit:
scripts: missing brackets in kw.sh

Due to missing brackets, kw.sh script may globe PLATFORM_TYPE variable.
It's not critical issue, as it may happened during "echo" call. This
issue is noticed by shellcheck (SC2086)

Add double quotes around PLATFORM_TYPE variable to prevent possible
globing.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

Commit:
scripts: fix shellcheck issues in tools/functions.sh

During shellcheck scan, next issues were met in function.sh script:
Line 1:  SC2148 Missing shebang.
Line 8:  SC2034 and SC2154 rootdir variable referenced but not assigned.
Line 11: SC2086 missing double quote around $0 variable.

In order to fix issues:
Line 1: add shellcheck directive to state shell used.
Line 8: add error in case if rootdir variable is not set
Disable shellcheck SC2034 warning, as this variable should be set by
script.
Line 11: add double quotes arond $0.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

Commit:
scripts: fix shellcheck issues in entrypoint.sh

During shellcheck scan of entrypoint.sh next issues were found:
Line 35: SC2153 SC2086 - no check if "cd" successful, missing double
quotes around variable.

Add double quotes around INSTALL_DIR variable and exit 1 if directory
point by this variables doesn't exist.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

Commit:
scripts: fix shellcheck issues in clang-format.sh

During shellcheck scan of clang-format.sh issue raised:
Line 13: SC1091 - Due to wrong path in shellcheck directive, it was not
able to follow sourced script.

Set right path in shellcheck directive.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

Commit:
scripts: fix shellcheck issues in tests/test_gw_repeater.sh

During shellcheck scan of test_gw_repeater.sh next issues were found:
Line 104: SC2086 - no double quotes around GW_EXTRA_OPT variable.
Line 117: SC2086 - no double quotes around RP_EXTRA_OPT variable.

Pass RP_EXTRA_OPT and GW_EXTRA_OPT variables as bash arrays to run.sh
script to avoid extra empty string argument.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -1,9 +1,10 @@
 
-scriptdir="$(cd "${0%/*}" && pwd)"
-rootdir="${scriptdir%/*/*}"
+scriptdir=$(dirname "$(readlink -f "$0")")
```
Which shellcheck error are you solving here? The original is the typical way of getting the current directory, why change it?

MuteSpirit replied:
Reverted changes.

I found why we see different number of shellcheck issues:
1) Ubuntu 18.04 uses shellcheck 0.4.6
    To install newer version (0.7.0) it's needed to use snap
    `sudo snap install shellcheck`
2) I run shellcheck without args
    Now use next command:
         `shellcheck -P SCRIPTDIR -x <script-path>`
3) I reconfigured Syntastic plugin in my Vim to see the same results.
     I've added next in my `~/.vimrc`
          `let g:syntastic_sh_shellcheck_args = "-P SCRIPTDIR -x"`

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -1,9 +1,10 @@
+rootdir=$(realpath "$scriptdir/../..")
 
-# shellcheck source=../../tools/functions.sh
+# shellcheck source=tools/functions.sh
```
The shellcheck invocation that is used inside an IDE relies on a path that is relative to the script itself, not to the root directory.

Fortunately, you can simply use `shellcheck -P SCRIPTDIR` to allow paths relative to the script.

MuteSpirit replied:
Agree.
Fixed used shellcheck command.

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -1,9 +1,10 @@
 
-# shellcheck source=../../tools/functions.sh
+# shellcheck source=tools/functions.sh
+# shellcheck disable=SC1091
```
No no no - make sure you include functions.sh in your shellcheck invocation and this won't be needed.

MuteSpirit replied:
Reverted my changes.

arnout commented on tools/docker/image-pull.sh:
```diff
@@ -45,11 +46,7 @@ main() {
-                                        distro="$(echo $NAME | awk '{print tolower($0)}')"
-                                        echo "$distro:$VERSION_ID"
-                                    ); shift ;;
+            -n | --native)          NATIVE=true; shift; ;;
```
Which shellcheck error is this trying to fix? That should be part of the commit message, ideally.

If it is just a refactoring because you think it's cleaner that way, do that in a separate commit. That way, it's easier to see what is really the shellcheck fix.

However, I don't think anybody uses the native option - it's an idea I had a long time ago but it doesn't really work. So instead, just remove the option.

MuteSpirit replied:
Add issues founded by shellcheck before refactoring into commit messages for all fixed scripts.
Create separate commit with removing "--native" option.

arnout commented on tools/functions.sh:
```diff
@@ -1,14 +1,15 @@
+#!/bin/sh
```
This script is not meant to be executed, so [use a directive](https://github.com/koalaman/shellcheck/wiki/SC2148) `# shellcheck shell=sh`.

MuteSpirit replied:
Followed your advice

arnout commented on tools/functions.sh:
```diff
@@ -1,14 +1,15 @@
 ###############################################################
 
-installdir="${rootdir}/build/install"
+export installdir="${rootdir:?}/build/install"
```
We really don't want to export it. This is an artifact of it being sourced. So here we need an exception.

rootdir is expected to have been set by the calling script. Instead of giving it a default, add a condition that errors out if it is not set.

MuteSpirit replied:
1) about `installdir` - agree. I'll add exception
2) about `rootdir`
`${rootdir:?}` does not define default value for variable but check that variable was set

Usage example
Add file `test.sh`
```
#!/bin/sh
. tools/functions.sh
echo "Hi"
```
and try execute it
```
$ sh test.sh 
test.sh: 9: tools/functions.sh: rootdir: parameter not set or null
```

arnout replied:
Oh, I didn't know that. Excellent!

arnout commented on tools/docker/image-pull.sh:
```diff
@@ -6,13 +6,14 @@
 
-scriptdir="$(cd "${0%/*}"; pwd)"
-rootdir="${scriptdir%/*/*}"
+scriptdir=$(dirname "$(readlink -f "$0")")
```
Again, please use
```suggestion
scriptdir="$(cd "${0%/*}" && pwd)"
```

MuteSpirit replied:
Eventually i did not get `scriptdir` via
```
scriptdir=$(dirname "$(readlink -f "$0")")
```
but followed shellcheck advice instead
```
    In tools/docker/image-pull.sh line 9:
    scriptdir="$(cd "${0%/*}"; pwd)"
                 ^----------^ SC2164: Use 'cd ... || exit' or 'cd ... || return' in case cd fails.
    
    Did you mean:
    scriptdir="$(cd "${0%/*}" || exit; pwd)"
```
and used
```
scriptdir="$(cd "${0%/*}" || exit 1; pwd)"
```

arnout replied:
Well, I don't know why shellcheck advises that, because it's completely equivalent to
```
scriptdir="$(cd "${0%/*}" && pwd)"
```
(which is what we use in most other places that get it right).

Anyway, doesn't matter much.

arnout commented on tools/klocwork/kw.sh:
```diff
@@ -21,7 +21,7 @@ declare PLATFORM_TYPE
 PLATFORM_TYPE=$PRPLMESH_PLATFORM_TYPE
 
-echo platfrom identified: $PLATFORM_TYPE
+echo platfrom identified: "$PLATFORM_TYPE"
```
While you're at it, fix the typo :-)

MuteSpirit replied:
Fixed

arnout reviewed (changes requested):
I think this is enough to merge. Well, it still needs a gitlab-ci update of course. You can use the same approach as the flake8 check, using [pipelinecomponents/shellcheck](https://hub.docker.com/r/pipelinecomponents/shellcheck). I don't think it's needed to add something similar as flake8-diff, review will make sure that modified scripts are shellchecked.

Commit:
scripts: image-pull.sh: remove "--native" option support

Option "--native" of script "tools/docker/image-pull.sh" makes
it to use same base OS image as the running system. This option is no
longer used.

Remove this option and any related code.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

Commit:
scripts: fix shellcheck issues in image-pull.sh

During shellcheck scan of image-pull.sh next issues were found:
Line 9: SC2164 - no check if cd changed directory sucessful
Line 12: SC1090, SC2086 - no directive to specify sourced script for
shellcheck, missing quotes around variable
Line 15: SC2086 - missing quotes
Line 36: SC2006 - usage of backticks instead of $(..) notation
Line 38: SC2181 - inderect exit code check of getopts
Line 64: SC2043 - VERBOSE variable is not used
Line 67: SC2068 - missing quotes around array expansions

In oder to fix issues:
Line 9: exit if cd failed
Line 12: add shellcheck directive to source script
add double quotes around rootdir variable
Line 15: add double quotes around $0
Line 36,38: remove backticks and place getopt command in if to check
return code directly
Line 64: remove verbose options, as variables set by it are not used
Line 67: add quotes around main call

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

MuteSpirit commented:
> I think this is enough to merge. Well, it still needs a gitlab-ci update of course. You can use the same approach as the flake8 check, using [pipelinecomponents/shellcheck](https://hub.docker.com/r/pipelinecomponents/shellcheck). I don't think it's needed to add something similar as flake8-diff, review will make sure that modified scripts are shellchecked.

I suggest to fix shellcheck issues in all shell scripts before enable this check on GitLab CI, otherwise developers besides me will have to fix shellcheck issues on any script change.

tomereli commented on tools/klocwork/kw.sh:
```diff
@@ -21,7 +21,7 @@ declare PLATFORM_TYPE
 PLATFORM_TYPE=$PRPLMESH_PLATFORM_TYPE
 
-echo platfrom identified: $PLATFORM_TYPE
+echo platform identified: "$PLATFORM_TYPE"
```
The commit message is way too long - suggestion:

> scripts: kw.sh: fix shellcheck issue SC2086
>
> Fix typo + add Quoting to prevent word splitting / globbing.
>For more information:
>  https://www.shellcheck.net/wiki/SC2206 -- Quote to prevent word splitting/g...
>  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
>
> Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>

The same for all commit messages in this series.

rmelotte replied:
It looks like a big part of the commit message has nothing to do with the changes, so Tomer's comment is not a nitpick IMO, please address it.

abelog replied:
Commit messages are fixed.

tomereli commented on tools/functions.sh:
```diff
@@ -1,14 +1,16 @@
 # See LICENSE file for more details.
 ###############################################################
 
-installdir="${rootdir}/build/install"
```
Here as well, the commit message is just a copy of the shellcheck output. It is not helpful IMO, please explain why you disabled SC2034 in the commit message and what syntax you used to set `installdir` - like you did here - https://github.com/prplfoundation/prplMesh/pull/1123#discussion_r407085029
You should also mention the quoting that was missing.

tomereli commented on tools/docker/runner/entrypoint.sh:
```diff
@@ -32,5 +32,5 @@ run ip link set      dev wlan2    up
 run ip link set      dev br-lan   up
 
-cd ${INSTALL_DIR}
+cd "${INSTALL_DIR}" || exit 1
```
Same comment as for previous commit messages.

tomereli commented on tests/test_gw_repeater.sh:
```diff
@@ -55,8 +55,8 @@ check_wsl() {
     # when the container was created
-    GW_EXTRA_OPT="--expose ${GW_UCC_PORT} --publish 127.0.0.1::${GW_UCC_PORT} --ipaddr 0.0.0.0"
-    RP_EXTRA_OPT="--expose ${RP_UCC_PORT} --publish 127.0.0.1::${RP_UCC_PORT} --ipaddr 0.0.0.0"
+    GW_EXTRA_OPT=(--expose "${GW_UCC_PORT}" --publish 127.0.0.1::"${GW_UCC_PORT}" --ipaddr 0.0.0.0)
```
The commit message explains that the shellcheck advice is not taken, but it doesn't explain what you did instead.

arnout replied:
> The commit message explains that the shellcheck advice is not taken, but it doesn't explain what you did instead.

This is still the case. Also, in the end you *do* follow the shellcheck advice (use a bash array).

tomereli commented on tools/docker/image-pull.sh:
```diff
@@ -17,7 +17,6 @@ usage() {
     echo "      -h|--help - show this help menu"
     echo "      -v|--verbose - verbosity on"
     echo "      -b|--base-image - Base OS image to use (Dockerfile 'FROM')"
-    echo "      -n|--native - Use the same base OS image as the running system"
```
Please drop this commit or fix the --native option - why isn't it usable? if indeed there is a bug, it should be fixed rather than removed.

arnout replied:
> Please drop this commit or fix the --native option - why isn't it usable?

That was my suggestion. Obviously, the explanation should have been part of the commit message.

```
The --native option doesn't really work because:
- for starters, it only applies to Debian derivatives, because the Dockerfile uses APT;
- in addition, it only works with Ubuntu because the package names differ slightly;
- finally, it only works with Ubuntu 18.04 and a few versions very close to it, because of package incompatibilities.

In other words, we will anyway need a separate Dockerfile for any platform that is not Ubuntu 18.04 (or close to it, e.g. 18.10 does work). Therefore the --native option is next to useless.
```

Unless you know someone who actually uses this option, @tomereli ?

tomereli replied:
All of Intel users are working on Ubuntu18.04 LTS server, so we used to use the -n option all the time. I haven't used it in a while since I didn't need to, so I'm not strictly against removing it - but I prefer to keep it unless its too much work.

arnout replied:
On an Ubuntu 18.04 the -n option is a NOP, it will set IMAGE to its default value `ubuntu:18.04`.

tomereli replied:
So it was probably used in the beginning when we used Ubuntu19 in our HyperV VMs - in that case I guess we can remove this.

tomereli reviewed (approved):
I think some of the commit messages should be revisited. I also think the --native option should not be removed. Other than that LGTM, so approving, but please fix those before you merge this.

rmelotte commented on tools/docker/image-pull.sh:
```diff
@@ -6,18 +6,18 @@
 ###############################################################
 
-scriptdir="$(cd "${0%/*}"; pwd)"
+scriptdir="$(cd "${0%/*}" || exit 1; pwd)"
```
In other scripts we used this instead, it's more concise and the result is the same:

```suggestion
scriptdir="$(cd "${0%/*}" && pwd)"
```

Note that both solutions do silence the shellcheck issue, but do *not* exit the script if `cd` fails (but it's not something you have to fix IMO).

abelog replied:
If `scriptdir` is missing,  what is the point going on with script? It will just report missing functions of `functions.sh`.

arnout replied:
> If `scriptdir` is missing, what is the point going on with script?

Excellent argument to prefer `&&`! The problem with `exit 1` is that an unwary reader may think it will exit the script, while it actually just exists the subshell created by `$(...)`. To exit the script you have to do
```suggestion
scriptdir="$(cd "${0%/*}" && pwd)" || exit 1
```
or `set -e`.

rmelotte replied:
>The problem with exit 1 is that an unwary reader may think it will exit the script, while it actually just exists the subshell created by $(...)

Exactly. I now realize my comment was not clear enough: 
when I said "it's not something you have to fix IMO" what I meant was "you don't have to make the script exit on error" and **not** "you don't have to follow my suggestion at all".

rmelotte reviewed (approved):
LGTM overall, but please take Tomer's comments about commit messages into account.

arnout commented on tests/test_gw_repeater.sh:
```diff
@@ -55,8 +55,8 @@ check_wsl() {
-    GW_EXTRA_OPT="--expose ${GW_UCC_PORT} --publish 127.0.0.1::${GW_UCC_PORT} --ipaddr 0.0.0.0"
-    RP_EXTRA_OPT="--expose ${RP_UCC_PORT} --publish 127.0.0.1::${RP_UCC_PORT} --ipaddr 0.0.0.0"
+    GW_EXTRA_OPT=(--expose "${GW_UCC_PORT}" --publish 127.0.0.1::"${GW_UCC_PORT}" --ipaddr 0.0.0.0)
+    RP_EXTRA_OPT=(--expose "${RP_UCC_PORT}" --publish 127.0.0.1::"${RP_UCC_PORT}" --ipaddr 0.0.0.0)
```
(nitpick)
```suggestion
    RP_EXTRA_OPT=(--expose "${RP_UCC_PORT}" --publish "127.0.0.1::${RP_UCC_PORT}" --ipaddr 0.0.0.0)
```
(I find it clearer to put the quotes on the outside.)

arnout commented on tests/test_gw_repeater.sh:
```diff
@@ -101,7 +101,7 @@ main() {
     [ "$START_GATEWAY" = "true" ] && {
         status "Start GW (Controller + local Agent)"
-        "${rootdir}"/tools/docker/run.sh -u "${UNIQUE_ID}" ${VERBOSE_OPT} ${FORCE_OPT} ${GW_EXTRA_OPT} \
+        "${rootdir}"/tools/docker/run.sh -u "${UNIQUE_ID}" ${VERBOSE_OPT} ${FORCE_OPT} "${GW_EXTRA_OPT[@]}" \
```
Doesn't this still trigger a shellcheck error for VERBOSE_OPT?

arnout commented on tools/docker/image-pull.sh:
```diff
@@ -17,7 +17,6 @@ usage() {
     echo "      -h|--help - show this help menu"
     echo "      -v|--verbose - verbosity on"
     echo "      -b|--base-image - Base OS image to use (Dockerfile 'FROM')"
-    echo "      -n|--native - Use the same base OS image as the running system"
```
Please update the commit message with the full explanation as mentioned in an earlier comment:
```
The --native option doesn't really work because:
- for starters, it only applies to Debian derivatives, because the Dockerfile uses APT;
- in addition, it only works with Ubuntu because the package names differ slightly;
- finally, it only works with Ubuntu 18.04 and a few versions very close to it, because of package incompatibilities.

In other words, we will anyway need a separate Dockerfile for any platform that is not Ubuntu 18.04 (or close to it, e.g. 18.10 does work). Therefore the --native option is next to useless.
```

abelog replied:
Done!

arnout commented on tools/docker/image-pull.sh:
```diff
@@ -33,15 +34,16 @@ pull_and_tag() {
 
     while true; do
         case "$1" in
-            -v | --verbose)         VERBOSE=true; shift ;;
```
The removal of the verbose option is not mentioned in the commit message. It's probably OK to do it (I don't think it has any effect anyway), but in general we prefer such changes to be in a separate commit.

Don't bother doing it now though, just for future.

arnout reviewed (approved):
There are still a few comments from other reviewers that haven't been resolved. Nothing major though.

abelog committed:
scripts: resove SC1091 for upload_owncloud.sh

Due to recent changes, functions.sh script now requires to set a
"rootdir" variable. Missing set of this variable provoked fail of
upload_owncloud.sh script.
Also, wrong pass in shellcheck include directive provoker SC1091
shellcheck error.

Set "rootdir" variable. Set right pass for shellcheck directive, so it
will follow sourced script.

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>
